File: agents/linux/testanyware_agent/query_resolver.py
"""Resolve an element query against a flat or nested element list.

Mirrors the query resolution logic from macOS (QueryResolver.swift) and
Windows (UiaQueryResolver.cs): filter by role, label, id, then disambiguate
by index.
"""

from __future__ import annotations

from testanyware_agent.models import ElementInfo


def resolve(
    elements: list[ElementInfo],
    role: str | None = None,
    label: str | None = None,
    element_id: str | None = None,
    index: int | None = None,
) -> tuple[str, ElementInfo | None, list[ElementInfo] | None]:
    """Resolve a query to a single element.

    Returns a tuple of (result_type, element, matches) where:
      - ("found", element, None) — single match
      - ("not_found", None, None) — no matches
      - ("multiple", None, matches) — ambiguous matches
    """
    flat = _flatten(elements)

    # Filter by role
    if role:
        flat = [e for e in flat if e.role == role]

    # Filter by label (substring, case-insensitive)
    if label:
        label_lower = label.lower()
        flat = [e for e in flat if e.label and label_lower in e.label.lower()]

    # Filter by id (exact match)
    if element_id:
        flat = [e for e in flat if e.id == element_id]

    if not flat:
        return ("not_found", None, None)

    # Disambiguate by index
    if index is not None:
        if 0 <= index < len(flat):
            return ("found", flat[index], None)
        return ("not_found", None, None)

    if len(flat) == 1:
        return ("found", flat[0], None)

    return ("multiple", None, flat[:10])


def _flatten(elements: list[ElementInfo]) -> list[ElementInfo]:
    """Flatten a nested element tree into a flat list."""
    result: list[ElementInfo] = []
    for element in elements:
        result.append(element)
        if element.children:
            result.extend(_flatten(element.children))
    return result

File: agents/macos/Sources/TestAnywareAgent/QueryResolver.swift
import TestAnywareAgentProtocol

/// The result of a query resolution attempt.
public enum QueryResult: Sendable {
    /// Exactly one element matched (or index disambiguation resolved to one).
    case found(ElementInfo)
    /// Multiple elements matched and no index was provided to disambiguate.
    case multiple([ElementInfo])
    /// No elements matched the query criteria.
    case notFound
}

/// Resolves element queries against a pre-walked `[ElementInfo]` tree.
public struct QueryResolver {

    /// Search `elements` (and their descendants) for elements matching the given criteria.
    ///
    /// - Parameters:
    ///   - elements: The top-level elements to search (TreeWalker output, or a window's children).
    ///   - role: If provided, only elements with this exact role are considered.
    ///   - label: If provided, only elements whose label contains this string (case-insensitive) are considered.
    ///   - id: If provided, only elements whose `id` exactly equals this string are considered.
    ///   - index: 1-based index for disambiguation when multiple elements match. If the Nth match
    ///     exists it is returned as `.found`; if out of range, `.notFound` is returned.
    /// - Returns: A `QueryResult` describing whether zero, one, or multiple elements matched.
    public static func resolve(
        in elements: [ElementInfo],
        role: UnifiedRole?,
        label: String?,
        id: String?,
        index: Int?
    ) -> QueryResult {
        var matches: [ElementInfo] = []
        collectMatches(in: elements, role: role, label: label, id: id, into: &matches)
        // The macOS AX tree exposes the same element through multiple
        // parent paths (notably NSStackView descendants and scroll-area
        // subtrees), so the same logical element can land in `matches`
        // more than once. Dedup on identifying attributes; element
        // identity in AX is structural, so two ElementInfos with the
        // same role/label/id/position/size are the same element.
        matches = dedupedByIdentity(matches)

        switch matches.count {
        case 0:
            return .notFound
        case 1:
            if let index {
                return index == 1 ? .found(matches[0]) : .notFound
            }
            return .found(matches[0])
        default:
            if let index {
                guard index >= 1 && index <= matches.count else { return .notFound }
                return .found(matches[index - 1])
            }
            return .multiple(matches)
        }
    }

    // MARK: - Private

    private static func collectMatches(
        in elements: [ElementInfo],
        role: UnifiedRole?,
        label: String?,
        id: String?,
        into matches: inout [ElementInfo]
    ) {
        for element in elements {
            if elementMatches(element, role: role, label: label, id: id) {
                matches.append(element)
            }
            if let children = element.children {
                collectMatches(in: children, role: role, label: label, id: id, into: &matches)
            }
        }
    }

    /// Collapses elements that share an identity-key (role, label, id,
    /// platformRole, position, size). Two `ElementInfo` values with the
    /// same key represent the same underlying AX element reached via
    /// different parent paths. Only elements with both `position` and
    /// `size` participate in dedup — without coordinates we cannot tell
    /// "same element via two paths" apart from "two distinct elements
    /// that happen to share role/label/id".
    static func dedupedByIdentity(_ elements: [ElementInfo]) -> [ElementInfo] {
        var seen: Set<String> = []
        var out: [ElementInfo] = []
        out.reserveCapacity(elements.count)
        for e in elements {
            guard let key = identityKey(e) else {
                out.append(e)
                continue
            }
            if seen.insert(key).inserted {
                out.append(e)
            }
        }
        return out
    }

    private static func identityKey(_ e: ElementInfo) -> String? {
        guard let pos = e.position, let sz = e.size else { return nil }
        return [
            e.role.rawValue,
            e.label ?? "_",
            e.id ?? "_",
            e.platformRole ?? "_",
            "\(pos.x),\(pos.y)",
            "\(sz.width),\(sz.height)",
        ].joined(separator: "|")
    }

    private static func elementMatches(
        _ element: ElementInfo,
        role: UnifiedRole?,
        label: String?,
        id: String?
    ) -> Bool {
        if let role, element.role != role {
            return false
        }
        if let label {
            guard let elementLabel = element.label,
                  elementLabel.localizedCaseInsensitiveContains(label) else {
                return false
            }
        }
        if let id, element.id != id {
            return false
        }
        return true
    }
}
